### PR TITLE
Upload logs when doing both build + deploy

### PR DESCRIPTION
**Pull request by CameronDiver**
Depends on https://github.com/resin-io/resin-builder/pull/146

### CODE REVIEW

File: lib/actions/deploy.coffee [hunk 1]
```diff
@@ -123,34 +142,65 @@ module.exports =
...
+				Promise.join(
+					logs
+					token
+					url
+					buildId
+					username
+					params.appName
+					uploadLogs
+				)
+				.then ->
```
Comment (Page-) [inline lib/actions/deploy.coffee]: This can be
```coffee
.return('Successfully uploaded logs')

File: lib/utils/docker.coffee [hunk 2]
```diff
@@ -128,7 +130,7 @@ exports.runBuild = (params, options, getBundleInfo) ->
 				buildSuccess: (image) ->
 					if options.tag?
 						console.log("Tagging image as #{options.tag}")
-					resolve(image)
+					resolve([image, logs])
```
Comment (Page-) [inline lib/utils/docker.coffee]: I think it'd be nicer to use `{ image, logs }` so that it's clear what each property means, rather than trying to map an index to a meaning (a similar idea to using `funcWithBoolOpts(true, false, false, true, false)` being much clearer with an object)

File: lib/actions/deploy.coffee [hunk 3]
```diff
@@ -112,6 +126,11 @@ module.exports =
 			parameter: 'source'
 			description: 'The source directory to use when building the image'
 			alias: 's'
+		},
+		{
+			signature: 'nologs'
```
Comment (Page-) [inline lib/actions/deploy.coffee]: I'm not convinced by the name - this one makes me think that it would hide local logs - maybe `no-logs-upload`?

File: lib/actions/deploy.coffee [hunk 4]
```diff
@@ -70,7 +84,7 @@ uploadToPromise = (request, size) ->
 			if obj.type?
 				switch obj.type
 					when 'error' then reject(new Error("Remote error: #{obj.error}"))
-					when 'success' then resolve(obj.image)
+					when 'success' then resolve([obj.image, obj.buildId])
```
Comment (Page-) [inline lib/actions/deploy.coffee]: I'm a big fan of objects for things like this (as mentioned in my other comment)

File: lib/actions/deploy.coffee [hunk 5]
```diff
@@ -58,6 +63,15 @@ performUpload = (image, token, username, url, size, appName) ->
...
 	uploadToPromise(post, size)
 
+uploadLogs = (logs, token, url, buildId, username, appName) ->
+	request = require('request')
+	url = url || process.env.RESINRC_RESIN_URL
+	request.post
+		url: getBuilderLogPushEndpoint(url, buildId, username, appName)
+		auth:
+			bearer: token
+		body: Buffer.from(logs).toString('base64')
```
Comment (Page-) [inline lib/actions/deploy.coffee]: Is the manual base64-ing necessary?

File: lib/actions/deploy.coffee [hunk 6]
```diff
@@ -6,6 +6,11 @@ getBuilderPushEndpoint = (baseUrl, owner, app) ->
 	escApp = encodeURIComponent(app)
 	"https://builder.#{baseUrl}/v1/push?owner=#{escOwner}&app=#{escApp}"
 
+getBuilderLogPushEndpoint = (baseUrl, buildId, owner, app) ->
+	escOwner = encodeURIComponent(owner)
+	escApp = encodeURIComponent(app)
+	"https://builder.#{baseUrl}/v1/pushLogs?owner=#{escOwner}&app=#{escApp}&buildId=#{buildId}"
```
Comment (Page-) [inline lib/actions/deploy.coffee]: It might be worth using `querystring.stringify({ owner, app, buildId })` - https://nodejs.org/api/querystring.html#querystring_querystring_stringify_obj_sep_eq_options

File: package.json [hunk 7]
```diff
@@ -73,6 +73,7 @@
     "resin-image-fs": "^2.1.2",
     "resin-image-manager": "^4.1.1",
     "resin-sdk-preconfigured": "^6.0.0",
+    "resin-settings-client": "^3.6.0",
```
Comment (Page-) [inline package.json]: This is unused?

File: lib/actions/deploy.coffee (hunk 5 above)
Comment (CameronDiver) [inline lib/actions/deploy.coffee]: What would be the alternative? AFAICT `request` doesn't do that itself, nor does a `Buffer`.

File: lib/actions/deploy.coffee (hunk 3 above)
Comment (CameronDiver) [inline lib/actions/deploy.coffee]: Good point, will change.

File: package.json (hunk 7 above)
Comment (CameronDiver) [inline package.json]: Yup, accidentally added it from an upcoming PR.

File: lib/actions/deploy.coffee [hunk 8]
```diff
@@ -58,6 +63,15 @@ performUpload = (image, token, username, url, size, appName) ->
 
 	uploadToPromise(post, size)
 
+uploadLogs = (logs, token, url, buildId, username, appName) ->
+	request = require('request')
+	url = url || process.env.RESINRC_RESIN_URL
```
Comment (emirotin) [inline lib/actions/deploy.coffee]: This is the wrong way to get this setting as it ignores the YAML configs. Look for resin-settings-client usages here and in resin-sdk-preconfigured.

File: lib/actions/deploy.coffee [hunk 9]
```diff
@@ -2,9 +2,14 @@ Promise = require('bluebird')
 dockerUtils = require('../utils/docker')
 
 getBuilderPushEndpoint = (baseUrl, owner, app) ->
-	escOwner = encodeURIComponent(owner)
-	escApp = encodeURIComponent(app)
-	"https://builder.#{baseUrl}/v1/push?owner=#{escOwner}&app=#{escApp}"
+	querystring = require('querystring')
```
Comment (emirotin) [inline lib/actions/deploy.coffee]: does it auto-escape?

Comment (CameronDiver) [inline lib/actions/deploy.coffee]: It does:
```
> querystring.stringify({ foo: 'Will "this" get "escaped?"' })
'foo=Will%20%22this%22%20get%20%22escaped%3F%22'
```

File: lib/actions/deploy.coffee (hunk 8 above)
Comment (CameronDiver) [inline lib/actions/deploy.coffee]: That's weird, `require('resin-settings-client').get('resinUrl')` is where the `url` value comes from, but I was experiencing weirdness where when setting the `RESINRC_RESIN_URL` env var to `resindev.io` that it wasn't being mirrored in the settings call. I've just re-tested and it is now, so will change.

File: lib/actions/deploy.coffee [hunk 10]
```diff
@@ -130,35 +148,69 @@ module.exports =
...
+			docker = dockerUtils.getDocker(options)
+			# Check input parameters
+			parseInput(params, options)
+			.then ([appName, build, source, imageName]) ->
+				tmpNameAsync()
+				.then (tmpPath) ->
+
+					# Setup the build args for how the build routine expects them
+					options = _.assign({}, options, { appName })
+					params = _.assign({}, params, { source: source })
```
Comment (emirotin) [inline lib/actions/deploy.coffee]: Why the change from the shortcut syntax?

Comment (CameronDiver) [inline lib/actions/deploy.coffee]: rebase hell... will change.

File: lib/actions/deploy.coffee [hunk 11]
```diff
@@ -61,23 +65,36 @@ performUpload = (image, token, username, url, size, appName, logStreams) ->
...
-			obj = JSON.parse(data)
-			if obj.type?
-				switch obj.type
-					when 'error' then reject(new Error("Remote error: #{obj.error}"))
-					when 'success' then resolve(obj.image)
-					when 'status' then logging.logInfo(logStreams, "Remote: #{obj.message}")
-					else reject(new Error("Received unexpected reply from remote: #{data}"))
-			else
-				reject(new Error("Received unexpected reply from remote: #{data}"))
+			Promise.try ->
```
Comment (Page-) [inline lib/actions/deploy.coffee]: I think this should be a normal try/catch, having this weird hybrid `Promise.try` inside a `new Promise` is really weird

File: lib/actions/deploy.coffee (hunk 5 above)
Comment (Page-) [inline lib/actions/deploy.coffee]: From https://github.com/request/request
> body - entity body for PATCH, POST and PUT requests. Must be a Buffer, String or ReadStream. If json is true, then body must be a JSON-serializable object.

That makes me think you can just pass the `Buffer` object directly

Comment (emirotin) [inline lib/actions/deploy.coffee]: I think it will be sent as octet-stream, not as base64-encoded, so the change on the server side will be needed as well.

Comment (CameronDiver) [inline lib/actions/deploy.coffee]: Yeah server-side changes will definitely be needed. I just deployed the current code for now, and I'll spend more time integrating the logs.
